internal/controller/gatewayclass.go

```diff
@@ -96,10 +103,27 @@ func (r *gatewayClassReconciler) hasMatchingController(obj client.Object) bool {
 func (r *gatewayClassReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
 	r.log.WithField("name", request.Name).Info("reconciling gatewayclass")
 
-	// Fetch the Gateway from the cache.
-	gc := &gatewayapi_v1alpha1.GatewayClass{}
-	if err := r.client.Get(ctx, request.NamespacedName, gc); api_errors.IsNotFound(err) {
-		// Not-found error, so trigger an OnDelete.
+	var gatewayClasses gatewayapi_v1alpha1.GatewayClassList
+	if err := r.client.List(context.Background(), &gatewayClasses); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error listing gatewayclasses: %w", err)
+	}
+
+	var controlledClasses controlledClasses
+
+	for i := range gatewayClasses.Items {
+		// avoid loop pointer issues
+		gc := gatewayClasses.Items[i]
+
+		if gc.Spec.Controller != r.controller {
+			// different controller, ignore.
+			continue
+		}
+
+		controlledClasses.add(&gc)
+	}
+
+	// no controlled gatewayclasses, trigger a delete
+	if controlledClasses.len() == 0 {
 		r.log.WithField("name", request.Name).Info("failed to find gatewayclass")
 
 		r.eventHandler.OnDelete(&gatewayapi_v1alpha1.GatewayClass{
```

Comment:
eventually we can drop passing the GC to the DAG, we really just need to update the status, but leaving it for now until I get a PR up for https://github.com/projectcontour/contour/issues/3864